packages/scratch-vm/scripts/codeGeneration/index.ts

```diff
@@ -1,41 +1,101 @@
 
- // Once you've update the files contents, write it back out to the guiIndexFile location
-}
\ No newline at end of file
+  /**
+   * Finds the indices of the code guards, splits the gui file into chunks such 
+   * that slices[0] is the code up to and including the first code guard, slices[1]
+   * is the code between the second code guard and the third code guard (including both),
+   * and slices[2] is the code from the last code guard (inclusive) to the end of the file
+   */
+  let lineArray : string[] = guiFileContent.split('\n');
+  const includesSubstr = (pat:string) => ((x:string) => { return x.includes(pat) });
+  const guards = [...iconImportGuards,...extensionInfoGuards];
+  const indices = guards.map(pat => lineArray.findIndex(includesSubstr(pat)));
+  let slices = [lineArray.slice(0,indices[0]+1),lineArray.slice(indices[1],indices[2]+1),lineArray.slice(indices[3])];
+
+  let extensionsAdded = 0;
+  let currGuiFileContent = guiFileContent;
+  const numExtensions = Object.keys(extensions).length;
+
+  for (const str in extensions) {
+    const ext = extensions[str];
+    const iconURL = ext.iconURL;
+    const insetIconURL = ext.insetIconURL;
+    const extensionId : string = str;
+
+    const relativePathToAssetsForExt = [...relativePathToAssetsFolder, extensionId];
+    const relativePathToExtension = [...relativePathToExtensionDir,extensionId];
+    const currPathToIconURL = path.resolve(__dirname,...relativePathToExtension,iconURL);
+    const currPathToInsetIconURL = path.resolve(__dirname,...relativePathToExtension,insetIconURL);
+    const newPathForIcons = path.resolve(__dirname, ...relativePathToAssetsForExt) + '/';
+
+    if (!existsSync(newPathForIcons)) mkdirSync(newPathForIcons);
+
+    //copy icons to new directory
+    copyFileSync(currPathToIconURL,newPathForIcons + iconURL);
+    copyFileSync(currPathToInsetIconURL,newPathForIcons + insetIconURL);
+
+    const iconURLName = extensionId + 'IconURL';
+    const insetIconURLName = extensionId + 'InsetIconURL';
+    const pathFromIndexJSXToAssets = `../../../extension-gallery-assets/${str}/`;
+    const iconImports = [`import ${iconURLName} from '${pathFromIndexJSXToAssets + iconURL}';`,
+                         `import ${insetIconURLName} from '${pathFromIndexJSXToAssets + insetIconURL}';`];
+
+    const menuItem = `    {
+      name: '${ext.title}',
+      extensionId: '${extensionId}',
+      iconURL: ${iconURLName},
+      insetIconURL: ${insetIconURLName},
+      description: '${ext.description}',
+      featured: true
+    },`;
+
+    /* First extension added: remove all code between code guards.
+     * All others: add to the code that's already been added between the
+     * code guards.
+     */
+    if (extensionsAdded !== 0) {
```

Comment:
Hmm, this system seems a little complicated (with the tallying of extensions and all this logic in the for loop).

I think it'd be possible to write a more obvious solution that looks like:
- Ingest the contents of the index.jsx file, store in lines array (same as current)
- remove all lines within the code gen guards
- Create to indexing variables like: `menuItemIndex` and `importIndex`
- Now loop over the extensions, adding the menu Item info at the `menuItemIndex` and the imports at the `importIndex`. Based on how many lines you just added, increment `menuItemIndex` and `importIndex`
- Finally, after exiting out of that for loop, write out the lines array to the file

Let me know if you'd like to discuss

Reply:
NOTE: This comment should have probably been posted at the top of the function..